File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <bits/stdc++.h>
using namespace std;

vector<int> DFS(vector<vector<int>>& graph, int s) {
	vector<bool> visited(graph.size(), false);
	vector<int> answer;
	stack<int> stack({ s });

	while (!stack.empty()) {
		int node = stack.top();
		stack.pop();
		if (!visited[node]) {
			visited[node] = true;
			answer.emplace_back(node);
			for (int i = 0; i < graph[node].size(); i++) {
				stack.push(graph[node][i]);
			}
		}
	}
	return answer;
}


vector<int> BFS(vector<vector<int>>& graph, int s) {
	vector<bool> visited(graph.size(), false);
	vector<int> answer;
	queue<int> queue({ s });

	while (!queue.empty()) {
		int node = queue.front();
		queue.pop();
		if (!visited[node]) {
			visited[node] = true;
			answer.emplace_back(node);
			for (int i = 0; i < graph[node].size(); i++) {
				queue.push(graph[node][i]);
			}
		}
	}
	return answer;
}

int main() {
	ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

	int v, e, s; cin >> v >> e >> s;
	vector<vector<int>> graph(v + 1);
	for (int i = 0; i < e; i++) {
		int a, b; cin >> a >> b;
		graph[a].push_back(b);
		graph[b].push_back(a);
	}

	for (int i = 1; i <= v; i++) sort(graph[i].rbegin(), graph[i].rend());
	vector<int> answer = DFS(graph, s);
	for (auto & i : answer) cout << i << " ";
	cout << "\n";

	for (int i = 1; i <= v; i++) sort(graph[i].begin(), graph[i].end());
	vector<int> answer2 = BFS(graph, s);
	for (auto & i : answer2) cout << i << " ";
	cout << "\n";

	return 0;
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from typing import List
from collections import deque
input = sys.stdin.readline

def DFS(graph: List, start: int) -> List:
    visited = [False for _ in range(len(graph))]
    stack, answer = [start], []
    while stack:
        node = stack.pop()
        if visited[node] == False:
            visited[node] = True
            answer.append(node)
            stack.extend(reversed(sorted(graph[node])))
    return answer


def BFS(graph: List, start: int) -> List:
    visited = [False for _ in range(len(graph))]
    queue, answer = deque(), []
    queue.append(start)
    while queue:
        node = queue.popleft()
        if visited[node] == False:
            visited[node] = True
            answer.append(node)
            queue.extend(sorted(graph[node]))
    return answer

v, e, s = map(int, input().split())
graph = [[] for _ in range(v+1)]

for i in range(e):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
    
print(*DFS(graph, s))
print(*BFS(graph, s))